sda/websocket.go

```diff
@@ -0,0 +1,97 @@
+package sda
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"time"
+
+	"github.com/dedis/cothority/log"
+	"github.com/dedis/cothority/network"
+	"golang.org/x/net/websocket"
+	"gopkg.in/tylerb/graceful.v1"
+)
+
+const readSize = 1024
+
+// WebSocket handles incoming client-requests using the WebService
+// protocol for JavaScript-compatibility.
+type WebSocket struct {
+	ServerIdentity *network.ServerIdentity
+	services       map[string]Service
+	server         *graceful.Server
+	mux            *http.ServeMux
+}
+
+// NewWebSocket opens a webservice-listener one port above the given
+// ServerIdentity.
+func NewWebSocket(si *network.ServerIdentity) *WebSocket {
+	w := &WebSocket{
+		ServerIdentity: si,
+		services:       make(map[string]Service),
+	}
+	webHost, err := getWebHost(w.ServerIdentity)
+	log.ErrFatal(err)
+	w.mux = http.NewServeMux()
+	w.server = &graceful.Server{
+		Timeout: 100 * time.Millisecond,
+		Server: &http.Server{
+			Addr:    webHost,
+			Handler: w.mux,
+		},
+	}
+	log.Lvl1("Starting to listen on", webHost)
+	go w.server.ListenAndServe()
+	return w
+}
+
+// RegisterService saves the service as being able to handle messages.
+func (w *WebSocket) RegisterService(service string, s Service) error {
+	w.services[service] = s
+	return nil
+}
+
+// Register a message-handler for a service.
+func (w *WebSocket) RegisterMessageHandler(service, path string) error {
+	log.Lvlf3("Registering websocket for ws://hostname/%s/%s", service, path)
+	h := func(ws *websocket.Conn) {
+		var buf []byte
+		for {
+			tmp := make([]byte, readSize)
+			n, err := ws.Read(tmp)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			buf = append(buf, tmp[:n]...)
+			if n < readSize {
+				break
+			}
+		}
+		reply, errCode := w.services[service].ProcessClientRequest(path, buf)
+		if errCode == 0 {
+			_, err := ws.Write(reply)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+		}
+		ws.WriteClose(errCode)
```

Comment:
Do we want that ? 
As a general rule, I've always practiced the "if you open it, you close it". Same story, in mobile environment, it's always better to keep things going I think. Why not returning the error code ?